nevergrad/optimization/optimizerlib.py

```diff
@@ -1109,4 +1164,74 @@ def __call__(self, instrumentation: Union[int, Instrumentation],
 MidQRBO = ParametrizedBO(initialization="Hammersley", middle_point=True).with_name("MidQRBO", register=True)
 LBO = ParametrizedBO(initialization="LHS").with_name("LBO", register=True)
 
+
+@registry.register
+class PBIL(base.Optimizer):
+    """
+    Implementation of the discrete algorithm PBIL
+
+    https://www.ri.cmu.edu/pub_files/pub1/baluja_shumeet_1994_2/baluja_shumeet_1994_2.pdf
+    """
+
+    # pylint: disable=too-many-instance-attributes
+
+    def __init__(self, instrumentation: Union[int, Instrumentation], budget: Optional[int] = None, num_workers: int = 1) -> None:
+        super().__init__(instrumentation, budget=budget, num_workers=num_workers)
+
+        num_categories = 2
+        self.p: np.ndarray = np.ones((1, self.dimension)) / num_categories
+        self.alpha = 0.3
+        self.llambda = max(100, num_workers)  # size of the population
+        self.mu = self.llambda // 2  # number of selected candidates
+        self._population: List[Tuple[float, np.ndarray]] = []
+
+    def _internal_ask_candidate(self) -> base.Candidate:
+        unif = self._rng.uniform(size=self.dimension)
+        data = (unif > self.p[0]).astype(float)
+        return self.create_candidate.from_data(data)
+
+    def _internal_tell_candidate(self, candidate: base.Candidate, value: float) -> None:
+        self._population.append((value, candidate.data))
+        if len(self._population) >= self.llambda:
+            self._population.sort(key=lambda tup: tup[0])
+            mean_pop: np.ndarray = np.mean([x[1] for x in self._population[: self.mu]])
+            self.p[0] = (1 - self.alpha) * self.p[0] + self.alpha * mean_pop
+            self._population = []
+
+
+@registry.register
+class cGA(base.Optimizer):
+    """
+    Implementation of the discrete cGA algorithm
+
+    https://pdfs.semanticscholar.org/4b0b/5733894ffc0b2968ddaab15d61751b87847a.pdf
+    """
+
+    # pylint: disable=too-many-instance-attributes
+
+    def __init__(self, instrumentation: Union[int, Instrumentation], budget: Optional[int] = None, num_workers: int = 1) -> None:
+        super().__init__(instrumentation, budget=budget, num_workers=num_workers)
+        num_categories = 2
+        self.p: np.ndarray = np.ones((1, self.dimension)) / num_categories
+        self.llambda = 2 * (self.budget if self.budget is not None else max(num_workers, 40))
+        self._value_candidate: Optional[Tuple[float, np.ndarray]] = None
+
+    def _internal_ask_candidate(self) -> base.Candidate:
+        unif = self._rng.uniform(size=self.dimension)
+        data = (unif > self.p[0]).astype(float)
```

Comment:
Ok p[0][i] is the probability of generating 0.
But below it looks like we increase p[0][i] if the winner has a 1. Did I miss something ?